rfc/0022-policy-lifecycle.md

```diff
@@ -0,0 +1,396 @@
+For instance if the leader fails to initialize a policy because of registry error, the following status condition will be set:
+
+```yaml
+conditions:
+  - type: Initialized
+    status: False
+    reason: PullError
+    message: "Cannot pull the policy from the registry."
+```
+
+### Rollback
+
+If a policy fails to load, the Policy Server will keep running with the previous valid policy.
+The user will be able to rollback to a previous policy version by updating the `Policy` CRD (via a spec field or an annotation, TBD).
+The controller will detect the change and use the `PolicyRevision` CRD to update the Policy CRD with the previous configuration.
+
+### Garbage Collection
+
+When a `Policy` is deleted, the controller will also delete all associated `PolicyRevisions`.
+Using the `OwnerReference` field is not feasible in this case, as the `PolicyRevision` may reside in a different namespace than the `AdmissionPolicy` or `AdmissionPolicyGroup`,
+or it could be versioning a `ClusterAdmissionPolicy` or `ClusterAdmissionPolicyGroup`, which are cluster-scoped resources.
+
+We will also need to implement a garbage collection mechanism to remove old precompiled modules from the shared cache.
+This will prevent the cache from growing indefinitely and consuming unnecessary resources.
+
+## Policy Lifecycle
+
+Given the concepts described above, the policy creation lifecycle will be as follows:
+
+1. The user creates a new `Policy`.
+2. The controller creates a new `PolicyRevision` resource with the policy content and sets the `enabled` field to `true`.
+   It also sets the `policyGeneration` field to the value of `metadata.generation` from the `Policy` CRD and adds the `Scheduled` status condition to both the `Policy` and `PolicyRevision` CRDs.
+3. The new `PolicyRevision` triggers the leader reconciler. The leader pulls the policy from the registry, precompiles the policy, and stores it in the shared cache.
+4. The leader validates the policy settings.
+5. The leader changes the status condition of type `Initialized` of the `PolicyRevision` to `True`.
+   - If an error occurred in the previous steps, the status condition of type `Initialized` will be set to `False` with the appropriate reason and message.
+6. The controller propagates the status condition to the `Policy` CRD.
+7. The replica reconciler of all the replicas including the leader loads the policy in the evaluation environment.
+8. Every replica will report the status of the policy to the `PolicyRevision` CRD, setting the status condition of type `Ready` to `True`.
+   - If an error occurs, the status condition of type `Ready` will be set to `False` with the appropriate reason and message.
+9. The `Ready` status is propagated to the `Policy` CRD.
+10. When all the replicas have set the `Ready` status to true, the controller creates a Webhook configuration with the `generation` query parameter set to the `PolicyRevision` generation.
```

Comment:
If we will update the webhooks only after all the replicas as ready, why do we need to have the generation in the path? On each case the old revision will be used? Is this necessary because the RFC expect to have a revision field in the `Policy` CRD?

Reply:
The goal is to continue serving the old policy until the new one is fully ready. Without including the generation in the path, any errors in loading the new policy by some replicas would prevent a proper rollout. In the current implementation, the policy remains marked as active because the old version was active, even if the new policy contains errors and the pod is in a CrashLoopBackoff state.